components/FlexibleHeader/src/private/MDCFlexibleHeaderMinMaxHeight.h

```diff
@@ -0,0 +1,147 @@
+// limitations under the License.
+
+#import <CoreGraphics/CoreGraphics.h>
+#import <Foundation/Foundation.h>
+
+@class MDCFlexibleHeaderTopSafeArea;
+@protocol MDCFlexibleHeaderMinMaxHeightDelegate;
+
+/**
+ Manages the height behavior for MDCFlexibleHeaderView.
+
+ In an ideal configuration, minMaxHeightIncludesSafeArea is disabled and the behavior of this class
+ is as follows.
+
+ minimumHeight and maximumHeight define the lower and upper bounds of MDCFlexibleHeaderView
+ without safe area insets taken into account. This class calculates minimum and maximum height
+ values that include the top safe area value and, most importantly, will react to changes of the top
+ safe area value.
+
+ Note: the top safe area value varies based on both the hardware (iPhone SE vs iPhone X) and the
+ hardware's orientation (iPhone X portrait, iPhone X landscape).
+
+ ## When minMaxHeightIncludesSafeArea is true
+
+ If minMaxHeightIncludesSafeArea is enabled then the behavior of this class is a bit more complex.
+ Notably, minimumHeight and maximumHeight are assumed to *include* the top safe area inset value.
+ This means that the client is expected to update the minimumHeight and maximumHeight any time the
+ top safe area value changes.
+
+ minMaxHeightIncludesSafeArea being set to true is the "legacy" behavior that existed prior to the
+ iPhone X and the existence of dynamic top safe areas. At the time of this property's introduction
+ we wanted to ensure that as many clients as possible would support the correct safe area behavior
+ with minimal code modifications. This meant assuming that minMaxHeightIncludesSafeArea was true but
+ finding a way for minimumHeight and maximumHeight to update in reaction to changes in the top safe
+ area.
+
+ If neither the minimumHeight nor maximumHeight are modified, this class will dynamically update
+ both properties as the top safe area inset changes with the assumption that it completely owns
+ both property's values. This covers the majority case of clients that are using the default
+ values. For all other clients that set explicit minimumHeight and maximumHeight values, we
+ recommend that they disable minMaxHeightIncludesSafeArea rather than attempt to implement the
+ correct top safe area logic themselves.
+ */
+__attribute__((objc_subclassing_restricted)) @interface MDCFlexibleHeaderMinMaxHeight : NSObject
+
+#pragma mark Initializing a min/max height object
+
+/**
+ Initializes this instance with a given top safe area instance.
+ */
+- (nonnull instancetype)initWithTopSafeArea:(nonnull MDCFlexibleHeaderTopSafeArea *)topSafeArea
+    NS_DESIGNATED_INITIALIZER;
+
+#pragma mark Configuring the minimum and maximum height
+
+/**
+ See MDCFlexibleHeaderView.h for complete documentation of this property.
+ */
+@property(nonatomic) CGFloat minimumHeight;
+
+/**
+ See MDCFlexibleHeaderView.h for complete documentation of this property.
+ */
+@property(nonatomic) CGFloat maximumHeight;
```

Comment:
suggestion: rename to maximum. Height is already in the class name and doesn't provide extra clarity in these properties..

Reply:
I'd like to keep these APIs matching the public APIs for clarity's sake.